File: 279_old.py
class Solution(object):
    def numSquares(self, n):
        """
        :type n: int
        :rtype: int
        """
        DP = [0] * (n+1)
        for i in xrange(1,n+1):
            j = 1
            while j*j <= i:
                rest = i - j*j
                val = DP[rest] + 1
                if DP[i] == 0 or DP[i] > val:
                    DP[i] = val
                j += 1
        
        return DP[n]

File: 279_old.cpp
int numSquares(int n) {
    int* DP = (int*) malloc(sizeof(int) * (n+1));
    for(int i=0;i<=n;i++) DP[i] = 0;
    for(int i=1;i<=n;i++){
        int j=1;
        while(j*j <=i){
            int rest = i - j*j;
            int val = DP[rest] + 1;
            if (DP[i] == 0 || DP[i] > val){
                DP[i] = val;
            }
            j += 1;
        }
    }
    int ans = DP[n];
    free(DP);
    return ans;
}